File: app/src/main/java/com/codemybrainsout/imageviewer/view/collection/CollectionPresenter.java
package com.codemybrainsout.imageviewer.view.collection;

import com.codemybrainsout.imageviewer.api.CollectionService;
import com.codemybrainsout.imageviewer.model.Collection;
import com.codemybrainsout.imageviewer.model.User;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Created by ahulr on 09-06-2017.
 */

public class CollectionPresenter implements CollectionContract.Presenter {

    private final CollectionContract.View mView;
    private CollectionService mCollectionService;

    public CollectionPresenter(CollectionService collectionService, CollectionContract.View view) {
        mView = view;
        mCollectionService = collectionService;
    }

    @Override
    public void openUser(User user) {
        mView.showUser(user);
    }

    @Override
    public void loadCollections(int page, int limit) {

        mView.showLoading();
        mCollectionService.getFeaturedCollections(page, limit)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Collection>>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull List<Collection> collections) {
                        mView.refreshCollections(collections);
                        mView.hideLoading();
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        e.printStackTrace();
                        mView.showError(e.getMessage());
                        mView.hideLoading();
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void loadMoreCollections(int page, int limit) {
        mCollectionService.getFeaturedCollections(page, limit)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Collection>>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull List<Collection> collections) {
                        mView.addCollections(collections);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        e.printStackTrace();
                        mView.showError(e.getMessage());
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void openCollection(Collection collection) {
        mView.showCollection(collection);
    }
}

File: app/src/main/java/com/codemybrainsout/imageviewer/view/collection/CollectionPresenter.kt
package com.codemybrainsout.imageviewer.view.collection

import com.codemybrainsout.imageviewer.api.CollectionService
import com.codemybrainsout.imageviewer.model.Collection
import com.codemybrainsout.imageviewer.model.User
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.annotations.NonNull
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by ahulr on 09-06-2017.
 */
class CollectionPresenter @Inject constructor(
    private val collectionService: CollectionService,
    private val mView: CollectionContract.View
) : CollectionContract.Presenter {

    override fun openUser(user: User?) {
        mView.showUser(user)
    }

    override fun loadCollections(page: Int, limit: Int) {
        mView.showLoading()
        collectionService.getFeaturedCollections(page, limit)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                mView.refreshCollections(it)
                mView.hideLoading()
            }, {
                it.printStackTrace()
                mView.showError(it.message)
                mView.hideLoading()
            })
    }

    override fun loadMoreCollections(page: Int, limit: Int) {
        collectionService.getFeaturedCollections(page, limit)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                mView.addCollections(it)
            }, {
                it.printStackTrace()
                mView.showError(it.message)
            })
    }

    override fun openCollection(collection: Collection?) {
        mView.showCollection(collection)
    }
}